src/main/scala/com/microsoft/hyperspace/index/rules/IndexFilter.scala

```diff
@@ -74,3 +80,151 @@ trait RankerIndexFilter extends IndexFilter with ActiveSparkSession {
+        index.unsetTagValue(plan, IS_HYBRIDSCAN_CANDIDATE)
+        index.setTagValue(plan, HYBRIDSCAN_RELATED_CONFIGS, curConfigs)
+      }
+    }
+  }
+
+  private def getHybridScanCandidate(
+      relation: FileBasedRelation,
+      index: IndexLogEntry,
+      hybridScanDeleteEnabled: Boolean): Option[IndexLogEntry] = {
+    // TODO: As in [[PlanSignatureProvider]], Source plan signature comparison is required to
+    //  support arbitrary source plans at index creation.
+    //  See https://github.com/microsoft/hyperspace/issues/158
+
+    val entry = relation.closestIndex(index)
+
+    val isHybridScanCandidate =
+      entry.withCachedTag(relation.plan, IndexLogEntryTags.IS_HYBRIDSCAN_CANDIDATE) {
+        // Find the number of common files between the source relation and index source files.
+        // The total size of common files are collected and tagged for candidate.
+        val (commonCnt, commonBytes) = relation.allFileInfos.foldLeft(0L, 0L) { (res, f) =>
+          if (entry.sourceFileInfoSet.contains(f)) {
+            (res._1 + 1, res._2 + f.size) // count, total bytes
+          } else {
+            res
+          }
+        }
+
+        val appendedBytesRatio = 1 - commonBytes / relation.allFileSizeInBytes.toFloat
+        val deletedBytesRatio = 1 - commonBytes / entry.sourceFilesSizeInBytes.toFloat
+
+        val deletedCnt = entry.sourceFileInfoSet.size - commonCnt
+        val isAppendAndDeleteCandidate = hybridScanDeleteEnabled && entry.hasLineageColumn &&
+          commonCnt > 0 &&
+          appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark) &&
+          deletedBytesRatio < HyperspaceConf.hybridScanDeletedRatioThreshold(spark)
+
+        // For append-only Hybrid Scan, deleted files are not allowed.
+        lazy val isAppendOnlyCandidate = deletedCnt == 0 && commonCnt > 0 &&
+          appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark)
+
+        val isCandidate = isAppendAndDeleteCandidate || isAppendOnlyCandidate
+        if (isCandidate) {
+          entry.setTagValue(
+            relation.plan,
+            IndexLogEntryTags.COMMON_SOURCE_SIZE_IN_BYTES,
+            commonBytes)
+
+          // If there is no change in source dataset, the index will be applied by
+          // RuleUtils.transformPlanToUseIndexOnlyScan.
+          entry.setTagValue(
+            relation.plan,
+            IndexLogEntryTags.HYBRIDSCAN_REQUIRED,
+            !(commonCnt == entry.sourceFileInfoSet.size
+              && commonCnt == relation.allFileInfos.size))
+        }
+        isCandidate
+      }
+    if (isHybridScanCandidate) {
+      Some(entry)
+    } else {
+      None
+    }
+  }
+
+  override def reason: String = {
+    "Underlying data is changed. Hybrid Scan Configs: \n" +
+      s"${IndexConstants.INDEX_HYBRID_SCAN_ENABLED}: ${HyperspaceConf.hybridScanEnabled(spark)}, " +
+      s"${IndexConstants.INDEX_HYBRID_SCAN_APPENDED_RATIO_THRESHOLD}: " +
```

Comment:
Hmm. It's possible that an index was not selected b/c of the threshold constraint was not met. How would this `reason` help in that case?

Reply:
Maybe, another way to achieve a better message is by using the tag (we can set the filter reason tag to the index log entry, for example)

Reply:
If we go with the log entry tagging approach, we should handle it now, no? e.g., the static `reason` doesn't make sense anymore.

Reply:
hmm if so we don't need to split a rule into multiple filters, thus I'd like to keep `reason` in somehow. 🤔

Reply:
Having multiple filters seems fine to me even without `reason`?

Reply:
But no explicit reason/condition for splitting them 🤔ok let's just remove it for now.

Reply:
I think the class name seems sufficient for the separation of concerns. Can you actually tag with the reason since this PR contains the actual filter implementations? I would like to confirm whether this approach makes sense.

Reply:
@imback82 updated PR - I'll add some tests for that msg but please have a look

Reply:
I took a quick look at it and the new approach looks fine to me. I will do another round of review. Thanks!